File: python/tests/pressure/lps33hw_test_unit.py
import sys

from periph.connection.i2c_mock import I2CConnectionMock
from periph.chips.pressure.lps33hw import LPS33HWFull

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


def preload_identity(connection):
    """Preload registers so the chip responds to its expected defaults and
    returns sensible pressure/temperature values from the burst read."""
    connection.set_register(LPS33HWFull._REG_WHO_AM_I, 0xB1)
    connection.set_register(LPS33HWFull._REG_STATUS, 0x03)
    # PRESS_OUT_XL..TEMP_OUT_H (5 bytes starting at 0x28):
    # raw_press = 0x100000 = 1048576 -> 1048576 * 100 / 4096 = 25600 Pa = 256 hPa
    # raw_temp  = 0x0A00 = 2560 -> 25.60 C
    connection.set_register(LPS33HWFull._REG_PRESS_XL, 0x00, 0x00, 0x10, 0x00, 0x0A)


connection = I2CConnectionMock()
preload_identity(connection)

sensor = LPS33HWFull(connection)
check_true('init_who_am_i_check', True)

# Verify the three init writes: SWRESET, default CTRL_REG2, default CTRL_REG1
ctrl2_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_CTRL_REG2]
check_true('init_writes_swreset', len(ctrl2_writes) >= 1 and ctrl2_writes[0][1] == 0x04)
check_true('init_writes_ctrl_reg2_default', len(ctrl2_writes) >= 2 and ctrl2_writes[1][1] == 0x10)
ctrl1_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_CTRL_REG1]
check_true('init_writes_ctrl_reg1_default', ctrl1_writes[-1][1] == 0x12)

# pressure(): 1048576 * 100 / 4096 = 25600.0 Pa
check_true('pressure', abs(sensor.pressure() - 25600.0) < 1e-6)
# temperature(): 2560 / 100 = 25.60 C
check_true('temperature', abs(sensor.temperature() - 25.60) < 1e-6)

# status(): STATUS = 0x03 -> p_da=1, t_da=1, p_or=0, t_or=0
check_true('status', sensor.status() == {'p_da': True, 't_da': True, 'p_or': False, 't_or': False})

# configure(odr=2, bdu=True, en_lpfp=True, lpfp_cfg=1): CTRL_REG1 =
# (2<<4)|(1<<3)|(1<<2)|(1<<1)|0 = 0x20|0x08|0x04|0x02 = 0x2E
sensor.configure(odr=2, bdu=True, en_lpfp=True, lpfp_cfg=1)
ctrl1_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_CTRL_REG1]
check_true('configure_writes_ctrl_reg1', ctrl1_writes[-1][1] == 0x2E)

# set_pressure_offset(2.5): raw = 2.5 * 16 = 40 -> RPDS_L=40, RPDS_H=0
sensor.set_pressure_offset(2.5)
rpds_l_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_RPDS_L]
rpds_h_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_RPDS_H]
check_true('set_pressure_offset_rpds_l', rpds_l_writes[-1][1] == 40)
check_true('set_pressure_offset_rpds_h', rpds_h_writes[-1][1] == 0)

# set_pressure_offset(-1.0): raw = -16 -> two's complement = 0xFFF0
sensor.set_pressure_offset(-1.0)
rpds_l_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_RPDS_L]
rpds_h_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_RPDS_H]
check_true('set_pressure_offset_negative_rpds_l', rpds_l_writes[-1][1] == 0xF0)
check_true('set_pressure_offset_negative_rpds_h', rpds_h_writes[-1][1] == 0xFF)

# configure_pressure_interrupt(high_en=True, low_en=True, threshold_hPa=5.0,
# latch=True): raw_ths = 5.0 * 16 = 80 -> THS_P_L=80, THS_P_H=0; INTERRUPT_CFG
# keeps top nibble (currently 0) | 0x04 (LIR) | 0x02 (PLE) | 0x01 (PHE) = 0x07
sensor.configure_pressure_interrupt(high_en=True, low_en=True, threshold_hPa=5.0, latch=True)
ths_l_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_THS_P_L]
ths_h_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_THS_P_H]
check_true('configure_pressure_interrupt_ths_l', ths_l_writes[-1][1] == 80)
check_true('configure_pressure_interrupt_ths_h', ths_h_writes[-1][1] == 0)
ic_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_INTERRUPT_CFG]
check_true('configure_pressure_interrupt_cfg', ic_writes[-1][1] == 0x07)

# configure_interrupt(drdy=True, f_fth=False, f_ovr=False, f_fss5=False,
# int_s=3, active_low=True, open_drain=True): CTRL_REG3 =
# (1<<7)|(1<<6)|(0<<5)|(0<<4)|(0<<3)|(1<<2)|3 = 0xC4|0x04|0x03 = 0xC7
sensor.configure_interrupt(drdy=True, f_fth=False, f_ovr=False, f_fss5=False,
                          int_s=3, active_low=True, open_drain=True)
ctrl3_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_CTRL_REG3]
check_true('configure_interrupt_ctrl_reg3', ctrl3_writes[-1][1] == 0xC7)

# enable_fifo(mode=1, watermark=15): FIFO_CTRL = (1<<5)|15 = 0x2F; CTRL_REG2 |= 0x40
sensor.enable_fifo(mode=1, watermark=15)
fifo_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_FIFO_CTRL]
check_true('enable_fifo_writes_fifo_ctrl', fifo_writes[-1][1] == 0x2F)
ctrl2_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_CTRL_REG2]
check_true('enable_fifo_sets_fifo_en', ctrl2_writes[-1][1] & 0x40)

# disable_fifo: FIFO_EN cleared, FIFO_CTRL=0
sensor.disable_fifo()
ctrl2_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_CTRL_REG2]
fifo_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_FIFO_CTRL]
check_true('disable_fifo_clears_fifo_en', not (ctrl2_writes[-1][1] & 0x40))
check_true('disable_fifo_writes_zero', fifo_writes[-1][1] == 0)

# fifo_status(): FIFO_STATUS = 0xE2 -> fth=1, ovr=1, count=34 (clip to mask & 0x3F)
connection.set_register(LPS33HWFull._REG_FIFO_STATUS, 0xE2)
check_true('fifo_status',
           sensor.fifo_status() == {'fth': True, 'ovr': True, 'count': 0x22})

# read_fifo() with 2 samples preloaded. Each is 5 bytes.
connection2 = I2CConnectionMock()
preload_identity(connection2)
# Override FIFO_STATUS to claim 2 samples queued
connection2.set_register(LPS33HWFull._REG_WHO_AM_I, 0xB1)
connection2.set_register(LPS33HWFull._REG_STATUS, 0x03)
connection2.set_register(LPS33HWFull._REG_FIFO_STATUS, 0x02)
connection2.set_register(LPS33HWFull._REG_PRESS_XL,
                         0x00, 0x00, 0x10, 0x00, 0x0A,
                         0x00, 0x00, 0x10, 0x00, 0x0A)
sensor2 = LPS33HWFull(connection2)
samples = sensor2.read_fifo()
check_true('read_fifo_length', len(samples) == 2)
check_true('read_fifo_pressure', abs(samples[0][0] - 25600.0) < 1e-6)
check_true('read_fifo_temperature', abs(samples[0][1] - 25.60) < 1e-6)

# one_shot(): sets ONE_SHOT bit, returns (P, T) tuple
connection3 = I2CConnectionMock()
preload_identity(connection3)
connection3.set_register(LPS33HWFull._REG_WHO_AM_I, 0xB1)
connection3.set_register(LPS33HWFull._REG_STATUS, 0x03)
connection3.set_register(LPS33HWFull._REG_PRESS_XL, 0x00, 0x00, 0x10, 0x00, 0x0A)
sensor3 = LPS33HWFull(connection3)
p_pa, t_c = sensor3.one_shot()
check_true('one_shot_pressure', abs(p_pa - 25600.0) < 1e-6)
check_true('one_shot_temperature', abs(t_c - 25.60) < 1e-6)

# reset(): writes SWRESET then default CTRL_REG2/CTRL_REG1
sensor.reset()
ctrl2_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_CTRL_REG2]
ctrl1_writes = [w for w in connection.writes if len(w) == 2 and w[0] == LPS33HWFull._REG_CTRL_REG1]
check_true('reset_writes_swreset_last', ctrl2_writes[-2][1] == 0x04)
check_true('reset_restores_ctrl_reg2_default', ctrl2_writes[-1][1] == 0x10)
check_true('reset_restores_ctrl_reg1_default', ctrl1_writes[-1][1] == 0x12)

print('===DONE: {} passed, {} failed==='.format(passed, failed))
sys.exit(0 if failed == 0 else 1)

File: nodejs/tests/pressure/lps33hw_test_unit.js
'use strict';
const { I2CConnectionMock } = require('../../packages/periph/src/connection/i2c_mock');
const { LPS33HWMinimal, LPS33HWFull } = require('../../packages/periph/src/chips/pressure/lps33hw');

const _REG_INTERRUPT_CFG = 0x0B;
const _REG_THS_P_L       = 0x0C;
const _REG_THS_P_H       = 0x0D;
const _REG_WHO_AM_I      = 0x0F;
const _REG_CTRL_REG1     = 0x10;
const _REG_CTRL_REG2     = 0x11;
const _REG_CTRL_REG3     = 0x12;
const _REG_FIFO_CTRL     = 0x14;
const _REG_RPDS_L        = 0x18;
const _REG_RPDS_H        = 0x19;
const _REG_RES_CONF      = 0x1A;
const _REG_INT_SOURCE    = 0x25;
const _REG_FIFO_STATUS   = 0x26;
const _REG_STATUS        = 0x27;
const _REG_PRESS_XL      = 0x28;
const _REG_LPFP_RES      = 0x33;

let passed = 0;
let failed = 0;

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else { console.log('FAIL', label); failed++; }
}

function flushMicrotasks() {
    return new Promise((resolve) => setImmediate(resolve));
}

function preloadDataReady(connection) {
    connection.setRegister(_REG_STATUS, [0x03]);
}

function preloadPressureTemp(connection, pressurePa, temperatureC) {
    const rawP = (pressurePa * 4096.0) / 100.0;
    const rawT = (temperatureC * 100.0);
    const p24 = Math.round(rawP) & 0xFFFFFF;
    const t16 = Math.round(rawT) & 0xFFFF;
    connection.setRegister(_REG_PRESS_XL, [
        p24 & 0xFF,
        (p24 >> 8) & 0xFF,
        (p24 >> 16) & 0xFF,
        t16 & 0xFF,
        (t16 >> 8) & 0xFF,
    ]);
}

async function main() {
    const connection = new I2CConnectionMock();
    connection.setRegister(_REG_WHO_AM_I, [0xB1]);
    preloadDataReady(connection);
    preloadPressureTemp(connection, 101325.0, 25.0);

    const sensor = new LPS33HWMinimal(connection);
    await flushMicrotasks();
    checkTrue('init_chip_id_check', true);

    const initCtrl2 = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_CTRL_REG2);
    checkTrue('init_writes_ctrl_reg2', initCtrl2.length >= 2);
    const initCtrl1 = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_CTRL_REG1 && w[1] === 0x12);
    checkTrue('init_writes_ctrl_reg1_default', initCtrl1.length > 0);

    const t = await sensor.temperature();
    checkTrue('temperature', Math.abs(t - 25.0) < 1e-6);

    const p = await sensor.pressure();
    checkTrue('pressure', Math.abs(p - 101325.0) < 0.01);

    const cid = await sensor._readReg(_REG_WHO_AM_I, 1);
    checkTrue('chip_id', cid[0] === 0xB1);

    const full = new LPS33HWFull(connection);
    await flushMicrotasks();

    // configure(odr=2, bdu=1, enLpfp=1, lpfpCfg=1, lcEn=0, sim=0):
    // CTRL_REG1 = (2<<4)|(1<<3)|(1<<2)|(1<<1) = 0x2E.
    connection.setRegister(_REG_RES_CONF, [0x00]);
    await full.configure(2, 1, 1, 1, 0, 0);
    const ctrl1Write = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_CTRL_REG1 && w[1] === 0x2E).pop();
    checkTrue('configure_ctrl_reg1', ctrl1Write !== undefined);

    // oneShot: STATUS=0x03 (P_DA|T_DA) so the loop returns immediately.
    preloadDataReady(connection);
    preloadPressureTemp(connection, 101325.0, 25.0);
    const sample = await full.oneShot();
    checkTrue('one_shot_pressure', Math.abs(sample.pressure_Pa - 101325.0) < 0.01);
    checkTrue('one_shot_temperature', Math.abs(sample.temperature_C - 25.0) < 1e-6);
    const oneShotTrig = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_CTRL_REG2 && (w[1] & 0x01) === 0x01);
    checkTrue('one_shot_triggers_ctrl_reg2_bit0', oneShotTrig.length > 0);

    // status: STATUS register returns the raw byte.
    connection.setRegister(_REG_STATUS, [0x05]);
    checkTrue('status', (await full.status()) === 0x05);

    // setPressureOffset(offsetHPa=1.5): raw = round(1.5 * 16) = 24 -> 0x0018.
    await full.setPressureOffset(1.5);
    const rpdsLWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_RPDS_L && w[1] === 0x18).pop();
    const rpdsHWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_RPDS_H && w[1] === 0x00).pop();
    checkTrue('set_pressure_offset_rpds_l', rpdsLWrite !== undefined);
    checkTrue('set_pressure_offset_rpds_h', rpdsHWrite !== undefined);

    // setAutozero: INTERRUPT_CFG OR 0x20. Read-modify-write.
    connection.setRegister(_REG_INTERRUPT_CFG, [0x00]);
    await full.setAutozero();
    const autozeroWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_INTERRUPT_CFG && (w[1] & 0x20) === 0x20).pop();
    checkTrue('set_autozero', autozeroWrite !== undefined);

    // clearAutozero: INTERRUPT_CFG OR 0x10.
    connection.setRegister(_REG_INTERRUPT_CFG, [0x20]);
    await full.clearAutozero();
    const clearAutozeroWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_INTERRUPT_CFG && (w[1] & 0x10) === 0x10).pop();
    checkTrue('clear_autozero', clearAutozeroWrite !== undefined);

    // setAutorifp / clearAutorifp: 0x80 / 0x40 mask bits.
    connection.setRegister(_REG_INTERRUPT_CFG, [0x00]);
    await full.setAutorifp();
    const autorifpWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_INTERRUPT_CFG && (w[1] & 0x80) === 0x80).pop();
    checkTrue('set_autorifp', autorifpWrite !== undefined);
    connection.setRegister(_REG_INTERRUPT_CFG, [0x80]);
    await full.clearAutorifp();
    const clearAutorifpWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_INTERRUPT_CFG && (w[1] & 0x40) === 0x40).pop();
    checkTrue('clear_autorifp', clearAutorifpWrite !== undefined);

    // configureInterrupt(drdy=1, fFth=0, fOvr=0, fFss5=1, intS=3, activeLow=1, openDrain=0):
    // CTRL_REG3 = 0x80 | 0x20 | 0x04 | 0x03 = 0xA7.
    await full.configureInterrupt(1, 0, 0, 1, 3, 1, 0);
    const intCfgWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_CTRL_REG3 && w[1] === 0xA7).pop();
    checkTrue('configure_interrupt', intCfgWrite !== undefined);

    // configurePressureInterrupt(highEn=1, lowEn=0, thresholdHPa=2.0, latch=1):
    // rawThs = round(2.0 * 16) = 32 = 0x0020 -> THS_P_L=0x20, THS_P_H=0x00.
    // INTERRUPT_CFG lower nibble = (1<<2)|(1<<1) = 0x06 (DIFF_EN=latch, PHE=highEn).
    connection.setRegister(_REG_INTERRUPT_CFG, [0x00]);
    await full.configurePressureInterrupt(1, 0, 2.0, 1);
    const thsLWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_THS_P_L && w[1] === 0x20).pop();
    const thsHWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_THS_P_H && w[1] === 0x00).pop();
    const intCfgPressWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_INTERRUPT_CFG && (w[1] & 0x0F) === 0x06).pop();
    checkTrue('configure_pressure_interrupt_ths_l', thsLWrite !== undefined);
    checkTrue('configure_pressure_interrupt_ths_h', thsHWrite !== undefined);
    checkTrue('configure_pressure_interrupt_cfg', intCfgPressWrite !== undefined);

    // enableFifo(mode=1, watermark=10): FIFO_CTRL = (1<<5)|10 = 0x2A; CTRL_REG2 OR 0x40.
    connection.setRegister(_REG_CTRL_REG2, [0x00]);
    await full.enableFifo(1, 10);
    const fifoCtrlWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_FIFO_CTRL && w[1] === 0x2A).pop();
    const fifoEnableWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_CTRL_REG2 && (w[1] & 0x40) === 0x40).pop();
    checkTrue('enable_fifo_ctrl', fifoCtrlWrite !== undefined);
    checkTrue('enable_fifo_bit', fifoEnableWrite !== undefined);

    // disableFifo: CTRL_REG2 &= ~0x40, FIFO_CTRL=0.
    connection.setRegister(_REG_CTRL_REG2, [0x40]);
    await full.disableFifo();
    const fifoDisableWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_CTRL_REG2 && (w[1] & 0x40) === 0).pop();
    const fifoResetWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_FIFO_CTRL && w[1] === 0x00).pop();
    checkTrue('disable_fifo_bit', fifoDisableWrite !== undefined);
    checkTrue('disable_fifo_zero', fifoResetWrite !== undefined);

    // fifoStatus: raw FIFO_STATUS byte.
    connection.setRegister(_REG_FIFO_STATUS, [0x1F]);
    checkTrue('fifo_status', (await full.fifoStatus()) === 0x1F);

    // interruptStatus: raw INT_SOURCE byte.
    connection.setRegister(_REG_INT_SOURCE, [0x10]);
    checkTrue('interrupt_status', (await full.interruptStatus()) === 0x10);

    // resetLpf: single-byte read of LPFP_RES.
    const beforeLpfReads = connection.writes.filter((w) => w.length === 1 && w[0] === _REG_LPFP_RES).length;
    await full.resetLpf();
    const afterLpfReads = connection.writes.filter((w) => w.length === 1 && w[0] === _REG_LPFP_RES).length;
    checkTrue('reset_lpf_reads_lpfp_res', afterLpfReads > beforeLpfReads);

    // reset(): BOOT bit set then cleared; CTRL_REG2 writes 0x04 (RESET) then 0x10 (DEFAULT), CTRL_REG1 written.
    connection.setRegister(_REG_CTRL_REG2, [0x04]);
    await full.reset();
    const resetWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_CTRL_REG2 && w[1] === 0x04).pop();
    const defaultWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_CTRL_REG2 && w[1] === 0x10).pop();
    checkTrue('reset_writes_boot', resetWrite !== undefined);
    checkTrue('reset_writes_default', defaultWrite !== undefined);

    // reboot(): CTRL_REG2 = 0x80 (BOOT).
    connection.setRegister(_REG_INT_SOURCE, [0x00]);
    await full.reboot();
    const rebootWrite = connection.writes.filter((w) => w.length === 2 && w[0] === _REG_CTRL_REG2 && w[1] === 0x80).pop();
    checkTrue('reboot_writes_boot', rebootWrite !== undefined);

    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main();
